engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/HashCodeAnyNode.java

```diff
@@ -0,0 +1,258 @@
+package org.enso.interpreter.node.expression.builtin.meta;
+
+import com.oracle.truffle.api.CompilerDirectives.TruffleBoundary;
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.dsl.Fallback;
+import com.oracle.truffle.api.dsl.GenerateUncached;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.interop.InvalidArrayIndexException;
+import com.oracle.truffle.api.interop.UnsupportedMessageException;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.nodes.Node;
+import com.oracle.truffle.api.profiles.LoopConditionProfile;
+import com.oracle.truffle.api.profiles.ValueProfile;
+import java.time.LocalDateTime;
+import java.time.ZonedDateTime;
+import java.util.Arrays;
+import org.enso.interpreter.dsl.AcceptsError;
+import org.enso.interpreter.dsl.BuiltinMethod;
+import org.enso.interpreter.runtime.callable.UnresolvedConversion;
+import org.enso.interpreter.runtime.callable.UnresolvedSymbol;
+import org.enso.interpreter.runtime.error.WarningsLibrary;
+
+/**
+ * Hashing contract:
+ * Whenever two objects are equal ({@code EqualsAnyNode} returns {@code true}), their hashcode
+ * should equal.
+ * More formally:
+ * {@code For all objects o1, o2: if o1 == o2 then hash(o1) == hash(o2)}
+ */
+@BuiltinMethod(
+    type = "Any",
+    name = "hash_code",
+    description = "Returns has code of the object."
+)
+@GenerateUncached
+public abstract class HashCodeAnyNode extends Node {
+
+  public static HashCodeAnyNode build() {
+    return HashCodeAnyNodeGen.create();
+  }
+
+  public abstract long execute(@AcceptsError Object self);
+
+  /** Specializations for primitive values **/
+
+  @Specialization
+  long hashCodeForLong(long l) {
+    return l;
+  }
+
+  @Specialization
+  long hashCodeForInt(int i) {
+    return i;
+  }
+
+  @Specialization
+  long hashCodeForDouble(double d) {
+    // TODO: More precise conversion?
+    return Double.valueOf(d).longValue();
```

Comment:
[doubleToLongBits](https://docs.oracle.com/javase/8/docs/api/java/lang/Double.html#doubleToLongBits-double-) maybe?

Reply:
I'd suggest doing something similar to our UnorderedMultiValueKey.foldNumeric - if `d % 1.0 == 0.0` we should return the _hashcode_ of its `longValue` (not the value itself) and otherwise we should probably fallback to the Java's implementation of hashcode for Double, no?

Reply:
Seems reasonable. At the end I came up with:
```
return d % 1.0 == 0.0 ? (int) d : Double.hashCode(d);
```
This should ensure consistency.

Reply:
I'm not convinced.

We want `2 == 2.0`, so `2.hash_code == 2.0.hash_code`. How are you computing a hash code of an integer? Are you doing identity? `2.hash_code == 2`??? That does not sound like a good hashing function. Why not use `Integer.hashCode(i)` for that?

Reply:
1. Good question, I think you are right that for _int_ identity could be a valid hashing function. I'm still not sure if it is that good - I'm not sure how is hashmap allocating bins based on its hashes - my reasoning was that for example if I have 64 bins, I could get the last or first 6 bits of the hash and do my bins depending on that - in such case an identity hash is bad, because if I took the first 6 bits, the integers next to each other would land in the same bin. So if I do just integers close to each other, they could potentially land in the same bin and the hashmap will become O(N) instead of O(1) by having non-uniform distribution of objects into the bin. That's why I was convinced that a good hash is a 'high-entropy' one, where integers close to each other will get completely different hashcodes - much like in SHA1 etc. If the Java hashmap plays well with such 'monotonic' hashcodes, it should be OK.

Reply:
2. Still, that applies for `int`, but we are using at least `long`s and there we definitely want something else than `(int) longValue` which will just discard half of bits.

Reply:
3. I had a look in the `jshell` REPL and the results are concerning, I can see that for large values what we've been using in the Table library is still insufficient so far.

```
jshell> Double.valueOf(123e30+1).longValue()
$15 ==> 9223372036854775807

jshell> Double.valueOf(123e40+1).longValue()
$16 ==> 9223372036854775807

jshell> Double.valueOf(123e50+1).longValue()
$17 ==> 9223372036854775807

jshell> (123e50+1) % 1.0
$18 ==> 0.0

jshell> (123e50+1) % 1.0 == 0.0
$19 ==> true

jshell> Long.MAX_VALUE
$20 ==> 9223372036854775807
```

As you can see, for large double values, `Double.longValue` just returns `Long.MAX_VALUE`. Thus, any large integer represented as a double (i.e. values for which `x % 1.0 == 0.0`) would get a hashcode stemming from `Long.MAX_VALUE`, so _all_ of them will fall into the same bin in hashing. That seems rather bad. Of course it is probably not very common for applications to use such big integers stored as longs, but I don't think we should allow the performance of hashmaps to drop radically whenever it _is_ used.

I guess ideally we'll want to have a good interoperability with `BigInteger`s and will probably want to ensure that the hash of such big integers represented as doubles is consistent with them. If this is the goal right now, the right way may be to check if the integer fits into `long`, if yes then delegate to `Double.longValue(x).hashCode()`, otherwise to convert it into a BigInteger and get a hashcode of _that_.

Maybe however if we do not care as much about bigintegers right now (the equality of such big integers with double is wonky anyway as not all integers are representable in such ranges of doubles), we can do something else. My suggestion would be to do:

```
if (x % 1.0 == 0.0 && fitsInLong(x)) return Long.hashCode(Double.valueOf(x).longValue());
else return Double.hashCode(x);
```

i.e. if we can fit the value into a `long`, we want it to have the same hashcode as its `long` counterpart - so that `2 == 2.0` (on hashcodes). However, when it doesn't fit, we just want to have a good hashcode, so we delegate to `Double.hashCode` as for all non-integral doubles. 

This way we will still get a good distribution of hashes for all doubles.

Reply:
I am not sure I entirely understand your point. Nevertheless, how about keeping the specialization for double as
```
return d % 1.0 == 0.0 ? (int) d : Double.hashCode(d);
```
And make another specialization for `EnsoBigInteger`, which would just return `ensoBigInt.getValue.hashCode()`, where the hash code is properly handled by JDK, as `BigInteger` has `hashCode` overriden.

Another note about `HashCodeAnyNode` returning `long`: This is simply because we do not have `int` as a type registered in Truffle type system - only `long` (See https://github.com/enso-org/enso/blob/develop/engine/runtime/src/main/java/org/enso/interpreter/runtime/type/Types.java). Both `java.util.Map` and `org.graalvm.collections.EconomicMap` use `int` for hashing. Therefore, returning `long` that has the higher bits unused is on purpose.

Reply:
Why not the one I proposed?

I don't like this one for the reason that for doubles > MAX_INT it will hash _every_ such integral double to the same value - that is definitely not uniform at all.

If we want to have a specialization for big integer, we probably need to ensure that integral doubles have the same hashcode as the corresponding big integers.

We just _have_ to ensure that if `a == b` then also `a.hash_code == b.hash_code`, this is an absolute requirement for hashing. So if we want our equality to behave correctly on big integers and decimals:

```
> a = 123 * 10^40
>>> Nothing
> b = a + 1.0
>>> Nothing
> a
>>> 1230000000000000000000000000000000000000000
> b
>>> 1.23E42
> a == b
>>> true
```

we need to ensure that in the following example `b.hash_code` is also equal to `a.hash_code`.

So I think the specialization we need is:
```
long/int doDouble(double x) {
   if (x % 1.0 != 0.0) {
       return Double.hashCode(x);
   } else {
       if (fitsInLong(x)) { return Long.hashCode(Double.valueOf(x)); }
	   else { return EnsoBigInteger.fromDouble(x).hashCode(); }
   }
}
```

(I assume we have some `BigInteger.fromDouble` constructor)

Reply:
Should be fixed with https://github.com/enso-org/enso/pull/3878/commits/af6b29bc5c15690e2450a3318f72164c6a6795ab . I also added the tests that you suggested. Hope I managed to cover all the cases.